### ISSUE TITLE
pointer-compare check can be very expensive

### ISSUE THREAD

**Issue by marxin** [google/sanitizers#1324]
As reported in the GCC compiler:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=97229

right now `__asan::GetGlobalsForAddress` does a linear iteration over all globals. Which can be very slow. I suggest to use a better data structure (at least a sorted list of globals based on `addr`)? I've attached a patch to the upstream bug.

**Comment by kcc** [comment]
Thanks for the report. 
We do not consider this feature production-ready (nor it is enabled by default). 
It's unlikely that we'll have cycles to work on it in ASAN, sorry. 
Nice feature though, I'd love to see it working. 

Thinking about it a bit more... With top-byte-ignore (and HWASAN or memory tagging) 
this feature will be so much simpler, faster, and robust. 
https://github.com/google/sanitizers/blob/master/hwaddress-sanitizer/MTE-iSecCon-2018.pdf

On AArch64 we can already do it today in HWASAN. 
I don't think we can get to it any time soon, but volunteers are welcome to send patches. 
Basically, need to check that two pointers being compared have the same address tag.
